### PR TITLE
KAA-651: implementation

### PR DESCRIPTION
**Pull request by ashvayka**

### CODE REVIEW

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/configuration/base/AbstractConfigurationManager.java [hunk 1]
```diff
@@ -114,11 +117,20 @@ public ConfigurationHashContainer getConfigurationHashContainer() {
 
     private byte[] loadConfigurationData() {
         if (storage != null) {
-            LOG.debug("Loading configuration data from storage {}", storage);
-            try {
-                configurationData = toByteArray(storage.loadConfiguration());
-            } catch (IOException e) {
-                LOG.error("Failed to load configuration from storage", e);
+            if(state.isConfigurationVersionUpdated()){
```
Comment (ghost) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/configuration/base/AbstractConfigurationManager.java]: Code style spaces.

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/configuration/base/AbstractConfigurationManager.java [hunk 2]
```diff
@@ -114,11 +117,20 @@ public ConfigurationHashContainer getConfigurationHashContainer() {
...
+                LOG.debug("Clearing old configuration data from storage {}", storage);
```
Comment (ghost) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/configuration/base/AbstractConfigurationManager.java]: I think this trace should be as INFO level.

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/configuration/base/SimpleConfigurationStorage.java [hunk 3]
```diff
@@ -41,6 +41,16 @@ public SimpleConfigurationStorage(KaaClientPlatformContext context, String path)
         this.context = context;
         this.path = path;
     }
+    
+    @Override
+    public void clearConfiguration() throws IOException {
+        PersistentStorage storage = context.createPersistentStorage();
+        if (storage.exists(path)) {
```
Comment (ghost) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/configuration/base/SimpleConfigurationStorage.java:48]: Code style spaces.

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/configuration/base/SimpleConfigurationStorage.java [hunk 4]
```diff
@@ -41,6 +41,16 @@ public SimpleConfigurationStorage(KaaClientPlatformContext context, String path)
...
+            storage.delete(path);
+        }else{
+            LOG.trace("There is no configuration in clear yet");
```
Comment (ghost) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/configuration/base/SimpleConfigurationStorage.java]: Maybe "... to clear" instead if "in clear"?

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/configuration/base/AbstractConfigurationManager.java [hunk 5]
```diff
@@ -40,10 +41,12 @@
...
     private ConfigurationStorage storage;
     private ConfigurationHashContainer container;
+    private KaaClientState state;
 
-    public AbstractConfigurationManager(KaaClientProperties properties) {
+    public AbstractConfigurationManager(KaaClientProperties properties, KaaClientState state) {
         super();
         this.properties = properties;
+        this.state = state;
         container = new HashContainer();
```
Comment (ghost) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/configuration/base/AbstractConfigurationManager.java]: I suggest either add "this." or move up to a class member declaration.

File: client/client-multi/client-java-android/src/main/java/org/kaaproject/kaa/client/persistence/AndroidInternalPersistentStorage.java [hunk 6]
```diff
@@ -57,8 +57,12 @@ public boolean renameTo(String oldPath, String newPath) throws IOException {
...
     @Override
     public boolean exists(String path) {
-        File f = new File(context.getFilesDir(), path);
-        return f.exists();
+        return new File(context.getFilesDir(), path).exists();
+    }
+
+    @Override
+    public void delete(String path) throws IOException {
+        new File(context.getFilesDir(), path).delete();
```
Comment (shpigor) [inline client/client-multi/client-java-android/src/main/java/org/kaaproject/kaa/client/persistence/AndroidInternalPersistentStorage.java:65]: I think need to use Files.deleteIfExists(Paths.get(path)) method intead

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/FilePersistentStorage.java [hunk 7]
```diff
@@ -42,8 +41,12 @@ public OutputStream openForWrite(String path) throws IOException {
 
     @Override
     public boolean exists(String path) {
-        File f = new File(path);
-        return f.exists();
+        return new File(path).exists();
```
Comment (shpigor) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/FilePersistentStorage.java:44]: You can use Files.notExists() method

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/FilePersistentStorage.java [hunk 8]
```diff
@@ -27,8 +27,7 @@
 
     @Override
     public InputStream openForRead(String path) throws IOException {
-        File f = new File(path);
-        return new FileInputStream(f);
+        return new FileInputStream(new File(path));
```
Comment (shpigor) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/FilePersistentStorage.java:30]: You can use Files.newInputStream() method

File: client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/FilePersistentStorage.java [hunk 9]
```diff
@@ -42,8 +41,12 @@ public OutputStream openForWrite(String path) throws IOException {
...
     @Override
     public boolean exists(String path) {
-        File f = new File(path);
-        return f.exists();
+        return new File(path).exists();
+    }
+    
+    @Override
+    public void delete(String path) throws IOException {
+        new File(path).delete();
```
Comment (shpigor) [inline client/client-multi/client-java-core/src/main/java/org/kaaproject/kaa/client/persistence/FilePersistentStorage.java:49]: You can use Files.deleteIfExists() method

**Pull request merged**